UI improvements

pstojanovichtec opened the pull request.

pstojanovichtec committed:
UI improvements

pstojanovichtec committed:
1.3.0

vdumitraskovic commented on README.md:
```diff
@@ -53,16 +53,29 @@ interface DefinedRange {
+| `footer`           | `ReactNode`           | _optional_ | `undefined`       | custom footer component                                         |
+| `months`           | `string[]`            | _optional_ | `undefined`       | translations for months                                         |
+| `weekDays`         | `string[]`            | _optional_ | `undefined`       | translations for week days                                      |
+| `actions`          | `DefinedRanges`       | _optional_ | `undefined`       | translations for actins, and enabling/disabling default actions |
```
```suggestion
| `actions`          | `DefinedRanges`       | _optional_ | `undefined`       | translations for actions, and enabling/disabling default actions |
```

vdumitraskovic commented on src/components/DateRangePickerWrapper.tsx:
```diff
@@ -35,6 +39,12 @@ const DateRangePickerWrapper: React.FunctionComponent<DateRangePickerWrapperProp
 
+  const { setActions, setDaysOfWeek, setMonths } = useDateRangeContext();
+
+  if (props.months) setMonths(props.months);
```
Are we sure we want to trigger this on every render? Maybe we should wrap this into useEffect hook?

vdumitraskovic commented on src/components/Header.tsx:
```diff
@@ -30,96 +43,44 @@ interface HeaderProps {
 const Header: React.FunctionComponent<HeaderProps> = ({
   date,
-  setDate,
+  // setDate,
```
```suggestion
```

vdumitraskovic commented on src/components/Header.tsx:
```diff
@@ -30,96 +43,44 @@ interface HeaderProps {
+
+      {nextDisabled || (
+        <Grid item className={classes.iconContainer}>
+          <IconButton className={classes.icon} disabled={nextDisabled} onClick={onClickNext}>
```
If `nextDisabled` is true, the component will not be rendered, hence there is not need to disable the button.

vdumitraskovic commented on src/components/Header.tsx:
```diff
@@ -30,96 +43,44 @@ interface HeaderProps {
-  const handleYearChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
-    setDate(setYear(date, parseInt(event.target.value)));
-  };
+  const getMonth = () => MONTHS[parseOptionalDate(date, date).getMonth()];
```
Maybe it would be a bit better if we hide month and year in useDateRangeContext

vdumitraskovic commented on src/components/Header.tsx:
```diff
@@ -30,96 +43,44 @@ interface HeaderProps {
-      </Grid>
+      {prevDisabled || (
+        <Grid item className={classes.iconContainer}>
+          <IconButton className={classes.icon} disabled={prevDisabled} onClick={onClickPrevious}>
```
If `prevDisabled` is true, the component will not be rendered, hence there is not need to disable the button.

vdumitraskovic commented on src/consts.ts:
```diff
@@ -0,0 +1,11 @@
+export const WEEK_DAYS = ["Mo", "Tu", "We", "Th", "Fr", "Sa", "Su"];
+export const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sept", "Oct", "Nov", "Dec"];
+export const DATE_RANGE_ACTIONS = {
+  today: "Today",
+  yesterday: "Yesterday",
+  thisWeek: "This Week",
+  lastWeek: "Last Week",
+  last7Days: "Last 7 Days",
+  thisMonth: "This Month",
+  lastMonth: "Last Month",
+};
```
By adding `as const`, the typescript will do a better type inference. Without it the type would just be `string[]`.

```suggestion
export const WEEK_DAYS = ["Mo", "Tu", "We", "Th", "Fr", "Sa", "Su"] as const;
export const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sept", "Oct", "Nov", "Dec"] as const;
export const DATE_RANGE_ACTIONS = {
  today: "Today",
  yesterday: "Yesterday",
  thisWeek: "This Week",
  lastWeek: "Last Week",
  last7Days: "Last 7 Days",
  thisMonth: "This Month",
  lastMonth: "Last Month",
} as const;
```

vdumitraskovic commented on src/defaults.ts:
```diff
@@ -1,43 +1,57 @@
+      startDate: date,
+      endDate: date,
+    });
+
+  if (actions.yesterday)
+    actionList.push({
+      label: actions.yesterday,
+      startDate: addDays(date, -1),
+      endDate: addDays(date, -1),
+    });
+  if (actions.thisWeek)
+    actionList.push({
+      label: actions.thisWeek,
+      startDate: startOfWeek(date, { weekStartsOn: 1 }),
+      endDate: endOfWeek(date, { weekStartsOn: 1 }),
+    });
+  if (actions.lastWeek)
+    actionList.push({
+      label: actions.lastWeek,
+      startDate: startOfWeek(addWeeks(date, -1), { weekStartsOn: 1 }),
+      endDate: endOfWeek(addWeeks(date, -1), { weekStartsOn: 1 }),
+    });
+  if (actions.last7Days)
+    actionList.push({
+      label: actions.last7Days,
+      startDate: addWeeks(date, -1),
+      endDate: addDays(date, -1),
+    });
+  if (actions.thisMonth)
+    actionList.push({
+      label: actions.thisMonth,
+      startDate: startOfMonth(date),
+      endDate: endOfMonth(date),
+    });
+  if (actions.lastMonth)
+    actionList.push({
+      label: actions.lastMonth,
+      startDate: startOfMonth(addMonths(date, -1)),
+      endDate: endOfMonth(addMonths(date, -1)),
+    });
```
We don't want to do this calculation on every render. This is a good example where we should use `useMemo` hook.

pstojanovichtec committed:
Fixed:
- typo
- added "as const"
- added useMemo
- removed unused code
- extracted "getMonth" and "getYear" to "DateRangePickerContext"

vdumitraskovic commented on src/components/Header.tsx:
```diff
@@ -30,96 +43,38 @@ interface HeaderProps {
+      {prevDisabled || (
+        <Grid item className={classes.iconContainer}>
+          <IconButton className={classes.icon} onClick={onClickPrevious}>
+            <ChevronLeft color={prevDisabled ? "disabled" : "action"} />
```
If `prevDisabled` is true, the component will never be rendered. We can safely use "action" only.

vdumitraskovic commented on src/components/Header.tsx:
```diff
@@ -30,96 +43,38 @@ interface HeaderProps {
+      {nextDisabled || (
+        <Grid item className={classes.iconContainer}>
+          <IconButton className={classes.icon} onClick={onClickNext}>
+            <ChevronRight color={nextDisabled ? "disabled" : "action"} />
```
If `nextDisabled` is true, the component will never be rendered. We can safely use "action" only.
